### PR TITLE
Perform PhysicalMemoryLimit check for workstation GC, refactor GetLargestOnDieCacheSize into GetCacheSizePerLogicalCpu

### PR DESCRIPTION
**Pull request by tmds**
Some time ago I was looking into gc.cpp implementation of checking physical memory: https://github.com/dotnet/coreclr/issues/14991#issuecomment-348428003

This PR is making two changes:
- The code that checks GetPhysicalMemoryLimit is also used for workstation gc
-  Combine GCToOSInterface::GetLargestOnDieCacheSize and GetLogicalCpuCount into GetCacheSizePerLogicalCpu. These two functions are used together to determine how much cache there is per logical processor. Combining them avoids confusion what GetLogicalCpuCount is returning (e.g. physical processors on die, vs processors affinitized to process), GetLogicalCpuCount was already taking nr of cores into account to some extent and GetLogicalCpuCount returns 1 when not implemented (which may not be a proper default value).

CC @swgillespie @Maoni0

### CODE REVIEW

**Comment by benaadams** [comment]
Tizen jammed https://github.com/dotnet/coreclr/issues/15914

File: src/vm/util.cpp [hunk 1]
```diff
@@ -1854,9 +1854,10 @@ DWORD GetLogicalCpuCountFallback()
 
 #endif // _TARGET_X86_ || _TARGET_AMD64_
 
-size_t GetLargestOnDieCacheSize(BOOL bTrueSize)
+// fix this if/when AMD does multicore or SMT
```
Comment (swgillespie) [inline src/vm/util.cpp]: I guess this is a little outdated now... 😆

File: src/vm/util.cpp [hunk 2]
```diff
@@ -1911,6 +1912,31 @@ size_t GetLargestOnDieCacheSize(BOOL bTrueSize)
                         }   
                     }
 
+                    // TODO: Currently GetLogicalCpuCountFromOS() and GetLogicalCpuCountFallback() are broken on
```
Comment (swgillespie) [inline src/vm/util.cpp]: This whole codepath seems to be really outdated to me. At one point I asked around to see if these comments still apply, and to the knowledge of everyone I talked to the OS issues that this function is working around were fixed a long time ago.

We should probably rewrite this whole thing at some point...

Comment (Maoni0) [reply src/vm/util.cpp]: We should be using GetLogicalProcessorInformationEx on the OS versions where it's available and fall back to cpuid where it's not. but you don't have to do it in this PR; feel free to open an issue to track if you don't want to do it in this PR.

---
In reply to: [163342330](https://github.com/dotnet/coreclr/pull/15975#discussion_r163342330) [](ancestors = 163342330)

File: src/gc/gc.cpp [hunk 3]
```diff
@@ -35527,19 +35527,26 @@ size_t GCHeap::GetValidGen0MaxSize(size_t seg_size)
...
         dprintf (2, ("cache: %Id-%Id, cpu: %Id", 
-            GCToOSInterface::GetLargestOnDieCacheSize(FALSE),
-            GCToOSInterface::GetLargestOnDieCacheSize(TRUE),
-            GCToOSInterface::GetLogicalCpuCount()));
+            GCToOSInterface::GetCacheSizePerLogicalCpu(FALSE),
+            GCToOSInterface::GetCacheSizePerLogicalCpu(TRUE)));
+
+        int n_heaps = gc_heap::n_heaps;
+#else //SERVER_GC
+        size_t trueSize = GCToOSInterface::GetCacheSizePerLogicalCpu(TRUE);
```
Comment (tmds) [inline src/gc/gc.cpp]: @Maoni0 @swgillespie I wonder, for workstation gc, if we want to multiply here with GCToOSInterface::GetCurrentProcessCpuCount. wdyt?

Comment (swgillespie) [reply src/gc/gc.cpp]: I'll leave that to @Maoni0, I don't have any thoughts - not super familiar with this math.

Comment (Maoni0) [reply src/gc/gc.cpp]: we don't want to do that 'cause we don't assume that processes using workstation GC will own the machine. unless you specifically affinitize your process with fewer CPUs, GetCurrentProcessCpuCount will return all CPUs.

File: src/vm/util.cpp (hunk 2 above)
Comment (tmds) [reply src/vm/util.cpp]: I have created https://github.com/dotnet/coreclr/issues/16071.

**Comment by tmds** [comment]
@swgillespie @Maoni0 thanks for reviewing! Can one of you merge this PR?

**Comment by Maoni0** [comment]
thanks for doing this, @tmds!

**Pull request merged**